File: sinain-chat-agent/tools.py
"""Lean 7-tool surface for the sinain chat agent (validated in chat-harness-bench §9c).

KEEP set: sinain_memory_query, sinain_context (screen OCR/vision + audio + apps),
sinain_memory_store, read_file, bash (read-only-sandboxed), grep, glob. Each is wired as
a native OpenHands ToolDefinition whose executor calls sinain-core HTTP / the local
machine and emits contract events through a per-run sink.

Why so few: every tool's schema is system-prompt tokens on EVERY turn → slower TTFT.
Connectors (GSuite/Slack/Glean) come later via deferred/lazy MCP loading, not eager load.
"""
from __future__ import annotations

import glob as _glob
import json
import os
import subprocess
import urllib.parse
import urllib.request

from pydantic import BaseModel, Field, SecretStr  # noqa: F401 (SecretStr kept for parity; used by sidecar)

from openhands.sdk.llm.message import TextContent
from openhands.sdk.tool import Action, Observation, ToolAnnotations, ToolDefinition, ToolExecutor

CORE_URL = os.environ.get("SINAIN_CORE_URL", "http://localhost:9500")
_MAX = 4000  # cap tool output so a huge payload can't blow up TTFT (the sinain_context lesson)


# ── tool backends (HTTP to sinain-core + local read-only machine ops) ────────
def _get(path: str) -> str:
    try:
        with urllib.request.urlopen(CORE_URL + path, timeout=10) as r:
            return r.read().decode("utf-8", "replace")[:_MAX]
    except OSError as e:
        return f"[sinain-core unreachable at {CORE_URL}: {e}]"


def _post(path: str, body: dict) -> str:
    try:
        req = urllib.request.Request(CORE_URL + path, data=json.dumps(body).encode(),
                                     headers={"Content-Type": "application/json"}, method="POST")
        with urllib.request.urlopen(req, timeout=10) as r:
            return r.read().decode("utf-8", "replace")[:2000]
    except OSError as e:
        return f"[sinain-core unreachable at {CORE_URL}: {e}]"


_BANNED = ("rm ", "rm -", "mkfs", "dd ", ":(){", "shutdown", "reboot", "> /dev", "sudo ", "chmod ", "mv ")


def _bash(cmd: str) -> str:
    if any(b in cmd for b in _BANNED):
        return "[refused: potentially destructive command]"
    try:
        out = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=15)
        return (out.stdout + out.stderr)[:_MAX] or "[no output]"
    except subprocess.TimeoutExpired:
        return "[bash timeout 15s]"


def dispatch(name: str, args: dict) -> str:
    if name == "sinain_memory_query":
        # /knowledge/query (NOT /knowledge/facts): the facts endpoint only reads
        # an `entities` param and ignores `q`, so the query term never reached
        # the graph. /knowledge/query splits `q` into keywords and runs the
        # multi-store (local + workspace) lookup.
        return _get(f"/knowledge/query?q={urllib.parse.quote(args.get('query', ''))}&max={args.get('limit', 8)}")
    if name == "sinain_context":
        return (_get("/agent/digest") + "\n---\n" + _get("/agent/context"))[:_MAX]
    if name == "sinain_memory_store":
        # /knowledge/import expects entity/attribute/value triples: {"facts":[...]}
        # (the shape the MCP server's sinain_memory_store also uses). Sending
        # {"content": text} was rejected with "No 'facts' array found". Drop
        # empty/None fields and skip facts missing a required key.
        facts = []
        for f in args.get("facts", []):
            fact = {k: v for k, v in f.items() if v not in (None, "")}
            if fact.get("entity") and fact.get("attribute") and fact.get("value"):
                facts.append(fact)
        if not facts:
            return "[sinain_memory_store: each fact needs non-empty entity, attribute, and value]"
        return _post("/knowledge/import", {"facts": facts})
    if name == "read_file":
        try:
            with open(args["path"], encoding="utf-8") as fh:
                return fh.read()[:_MAX]
        except OSError as e:
            return f"[read_file error: {e}]"
    if name == "bash":
        return _bash(args.get("command", ""))
    if name == "grep":
        path = json.dumps(args.get("path", "."))
        return _bash(f"grep -rn -- {json.dumps(args.get('pattern', ''))} {path} | head -50")
    if name == "glob":
        return "\n".join(_glob.glob(args.get("pattern", ""), recursive=True)[:50]) or "[no matches]"
    return f"[unknown tool: {name}]"


# ── OpenHands ToolDefinition wiring ──────────────────────────────────────────
class MemoryQueryAction(Action):
    query: str
    limit: int = 8


class StoreFact(BaseModel):
    entity: str = Field(description="Entity the fact is about, kebab-case, e.g. 'sinain-hud' or 'igor'")
    attribute: str = Field(description="Attribute name, e.g. 'prefers', 'deadline', 'status'")
    value: str = Field(description="The fact content")
    confidence: float = Field(default=0.7, description="0..1 confidence, default 0.7")
    domain: str | None = Field(default=None, description="Optional domain tag, e.g. 'work', 'german'")


class MemoryStoreAction(Action):
    facts: list[StoreFact] = Field(description="One or more entity/attribute/value facts to store")


class BashAction(Action):
    command: str


class ReadFileAction(Action):
    path: str


class GrepAction(Action):
    pattern: str
    path: str = "."


class GlobAction(Action):
    pattern: str


class EmptyAction(Action):
    pass


class SinainObs(Observation):
    pass


# per-run event sink (set by the sidecar before each turn; chat is serialized per session)
SINK = None  # set to a callable emit(event_dict) by the sidecar


class _Exec(ToolExecutor):
    def __init__(self, tool_name: str) -> None:
        self.tool_name = tool_name

    def __call__(self, action: Action, conversation=None) -> Observation:  # noqa: ANN001
        args = action.model_dump()
        if SINK:
            SINK({"type": "tool_call", "tool_name": self.tool_name, "tool_args": args})
        result = dispatch(self.tool_name, args)
        if SINK:
            SINK({"type": "tool_result", "tool_name": self.tool_name, "tool_result": result[:500]})
        return SinainObs(content=[TextContent(text=result)])


def _annot() -> ToolAnnotations:
    return ToolAnnotations(readOnlyHint=True, destructiveHint=False, idempotentHint=False, openWorldHint=True)


def _make(name: str, action_cls: type, desc: str) -> type:
    def create(cls, conv_state=None, **params):  # noqa: ANN001, ARG001
        return [cls(description=desc, action_type=action_cls, observation_type=SinainObs,
                    executor=_Exec(name), annotations=_annot())]
    return type(f"SinainTool_{name}", (ToolDefinition,), {"create": classmethod(create)})


SPECS = [
    ("sinain_memory_query", MemoryQueryAction, "Query the user's knowledge graph (facts/entities)."),
    ("sinain_context", EmptyAction, "Get the user's current situation: digest + screen OCR/vision, audio, app history."),
    ("sinain_memory_store", MemoryStoreAction, "Store facts in the user's long-term memory (knowledge graph). Each fact is an entity/attribute/value triple; duplicates are deduped automatically."),
    ("read_file", ReadFileAction, "Read a UTF-8 text file by absolute path."),
    ("bash", BashAction, "Run a read-only shell command to orient on the machine."),
    ("grep", GrepAction, "Search file contents for a pattern under a path."),
    ("glob", GlobAction, "Find files matching a glob pattern."),
]


def register_all() -> list[str]:
    from openhands.sdk import register_tool
    for name, action_cls, desc in SPECS:
        register_tool(name, _make(name, action_cls, desc))
    return [n for n, _, _ in SPECS]

File: sinain-chat-agent-js/src/tools.ts
/** Lean 8-tool surface for the sinain chat agent — port of sinain-chat-agent/tools.py.
 *
 * KEEP set: sinain_memory_query, sinain_context, sinain_memory_episodes,
 * sinain_memory_store, read_file, bash (read-only-sandboxed), grep, glob.
 * Each executor calls sinain-core HTTP / the local machine and emits contract
 * events (tool_call / tool_result) through a per-turn sink, exactly like the
 * Python SINK pattern — so the stream-parsing side never needs tool plumbing.
 *
 * Why so few: every tool's schema is prompt tokens on EVERY turn → slower TTFT.
 * Connectors (GSuite/Slack/Glean) come later via deferred MCP loading.
 */
import { execFile } from "node:child_process";
import { readFile } from "node:fs/promises";
import { glob as fsGlob } from "node:fs/promises";
import { tool, type ToolSet } from "ai";
import { z } from "zod";

const CORE_URL = process.env.SINAIN_CORE_URL || "http://localhost:9500";
const MAX = 4000; // cap tool output so a huge payload can't blow up TTFT (the sinain_context lesson)

/** Per-turn event sink (set by the sidecar before each turn; chat is serialized). */
export type Sink = (ev: Record<string, unknown>) => void;
let sink: Sink | null = null;
export function setSink(s: Sink | null): void {
  sink = s;
}

async function get(path: string): Promise<string> {
  try {
    const res = await fetch(CORE_URL + path, { signal: AbortSignal.timeout(10_000) });
    return (await res.text()).slice(0, MAX);
  } catch (e) {
    return `[sinain-core unreachable at ${CORE_URL}: ${e instanceof Error ? e.message : e}]`;
  }
}

async function post(path: string, body: unknown): Promise<string> {
  try {
    const res = await fetch(CORE_URL + path, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(10_000),
    });
    return (await res.text()).slice(0, 2000);
  } catch (e) {
    return `[sinain-core unreachable at ${CORE_URL}: ${e instanceof Error ? e.message : e}]`;
  }
}

const BANNED = ["rm ", "rm -", "mkfs", "dd ", ":(){", "shutdown", "reboot", "> /dev", "sudo ", "chmod ", "mv "];

function bash(cmd: string): Promise<string> {
  if (BANNED.some((b) => cmd.includes(b))) {
    return Promise.resolve("[refused: potentially destructive command]");
  }
  return new Promise((resolve) => {
    execFile("/bin/bash", ["-c", cmd], { timeout: 15_000 }, (err, stdout, stderr) => {
      if (err && err.killed) return resolve("[bash timeout 15s]");
      resolve((stdout + stderr).slice(0, MAX) || "[no output]");
    });
  });
}

/** Wrap an executor so every call emits the contract's tool_call/tool_result events. */
function traced<A>(name: string, run: (args: A) => Promise<string>): (args: A) => Promise<string> {
  return async (args: A) => {
    sink?.({ type: "tool_call", tool_name: name, tool_args: args as Record<string, unknown> });
    const result = await run(args);
    sink?.({ type: "tool_result", tool_name: name, tool_result: result.slice(0, 500) });
    return result;
  };
}

export function buildTools(): ToolSet {
  return {
    sinain_memory_query: tool({
      description: "Query the user's knowledge graph (facts/entities).",
      inputSchema: z.object({
        query: z.string(),
        limit: z.number().int().optional().default(8),
      }),
      execute: traced("sinain_memory_query", async ({ query, limit }) =>
        // /knowledge/query (NOT /knowledge/facts): the facts endpoint ignores `q`.
        get(`/knowledge/query?q=${encodeURIComponent(query)}&max=${limit ?? 8}`)),
    }),
    sinain_context: tool({
      description: "Get the user's current situation: digest + screen OCR/vision, audio, app history.",
      inputSchema: z.object({}),
      execute: traced("sinain_context", async () =>
        (await get("/agent/digest") + "\n---\n" + await get("/agent/context")).slice(0, MAX)),
    }),
    sinain_memory_episodes: tool({
      description:
        "Recall past observed activity as dated episodes (conversations, meetings, work sessions). " +
        "Use for 'what happened earlier / summarize that meeting' — set include_text for raw excerpts.",
      inputSchema: z.object({
        query: z.string().optional().default(""),
        since: z.string().optional().default("").describe("ISO timestamp, e.g. 2026-07-03T10:00"),
        limit: z.number().int().optional().default(20),
        include_text: z.boolean().optional().default(false),
      }),
      execute: traced("sinain_memory_episodes", async ({ query, since, limit, include_text }) =>
        get(`/memory/episodes?q=${encodeURIComponent(query ?? "")}&since=${encodeURIComponent(since ?? "")}` +
            `&limit=${limit ?? 20}&text=${include_text ? "1" : "0"}`)),
    }),
    sinain_memory_store: tool({
      description: "Save a fact/note to the user's long-term memory.",
      inputSchema: z.object({ text: z.string() }),
      execute: traced("sinain_memory_store", async ({ text }) => {
        // /knowledge/import expects the sinain export shape {"facts":[{entity,
        // attribute,value}]} — free text becomes one note triple the
        // deterministic integrator dedups + makes queryable.
        const txt = (text ?? "").trim();
        if (!txt) return JSON.stringify({ ok: false, error: "nothing to store" });
        return post("/knowledge/import", {
          facts: [{ entity: "user", attribute: "note", value: txt, confidence: 0.7 }],
        });
      }),
    }),
    read_file: tool({
      description: "Read a UTF-8 text file by absolute path.",
      inputSchema: z.object({ path: z.string() }),
      execute: traced("read_file", async ({ path }) => {
        try {
          return (await readFile(path, "utf-8")).slice(0, MAX);
        } catch (e) {
          return `[read_file error: ${e instanceof Error ? e.message : e}]`;
        }
      }),
    }),
    bash: tool({
      description: "Run a read-only shell command to orient on the machine.",
      inputSchema: z.object({ command: z.string() }),
      execute: traced("bash", async ({ command }) => bash(command ?? "")),
    }),
    grep: tool({
      description: "Search file contents for a pattern under a path.",
      inputSchema: z.object({
        pattern: z.string(),
        path: z.string().optional().default("."),
      }),
      execute: traced("grep", async ({ pattern, path }) =>
        bash(`grep -rn -- ${JSON.stringify(pattern ?? "")} ${JSON.stringify(path ?? ".")} | head -50`)),
    }),
    glob: tool({
      description: "Find files matching a glob pattern.",
      inputSchema: z.object({ pattern: z.string() }),
      execute: traced("glob", async ({ pattern }) => {
        try {
          const out: string[] = [];
          for await (const p of fsGlob(pattern ?? "")) {
            out.push(String(p));
            if (out.length >= 50) break;
          }
          return out.join("\n") || "[no matches]";
        } catch (e) {
          return `[glob error: ${e instanceof Error ? e.message : e}]`;
        }
      }),
    }),
  };
}
